tests/subsys/bluetooth/mesh/scheduler_model/message_validity/src/main.c

```diff
@@ -0,0 +1,571 @@
+{
+}
+
+static void onoff_get(struct bt_mesh_onoff_srv *srv, struct bt_mesh_msg_ctx *ctx,
+		      struct bt_mesh_onoff_status *rsp)
+{
+}
+
+struct bt_mesh_onoff_srv_handlers onoff_srv_handlers = {
+	.set = onoff_set,
+	.get = onoff_get,
+};
+
+struct bt_mesh_onoff_srv onoff_srv = BT_MESH_ONOFF_SRV_INIT(&onoff_srv_handlers);
+
+static void srv_action_set_cb(struct bt_mesh_scheduler_srv *srv, struct bt_mesh_msg_ctx *ctx,
+			      uint8_t idx, struct bt_mesh_schedule_entry *entry)
+{
+	ztest_check_expected_value(srv->model);
+	ztest_check_expected_value(ctx);
+	ztest_check_expected_value(idx);
+	ztest_check_expected_data(entry, sizeof(struct bt_mesh_schedule_entry));
+}
+
+static struct bt_mesh_scheduler_srv sched_srv =
+	BT_MESH_SCHEDULER_SRV_INIT(&srv_action_set_cb, &time_srv);
+
+static void cli_action_status_handler(struct bt_mesh_scheduler_cli *cli,
+				      struct bt_mesh_msg_ctx *ctx, uint8_t idx,
+				      const struct bt_mesh_schedule_entry *action)
+{
+	ztest_check_expected_value(cli->model);
+	ztest_check_expected_value(ctx);
+	ztest_check_expected_value(idx);
+	ztest_check_expected_data(action, sizeof(struct bt_mesh_schedule_entry));
+}
+
+static void cli_status_handler(struct bt_mesh_scheduler_cli *cli, struct bt_mesh_msg_ctx *ctx,
+			       uint16_t schedules)
+{
+	ztest_check_expected_value(cli->model);
+	ztest_check_expected_value(ctx);
+	ztest_check_expected_value(schedules);
+}
+
+static struct bt_mesh_scheduler_cli sched_cli = {
+	.action_status_handler = &cli_action_status_handler,
+	.status_handler = &cli_status_handler,
+};
+
+/** Mocking mesh composition data
+ *
+ * The mocked composition data consists of 1 element with
+ * 1 Scheduler server, 1 Scheduler client, and 1 Generic onoff server.
+ *
+ * |      1     |
+ * |:----------:|
+ * | Sched. Srv |
+ * | Sched. Cli |
+ * | Onoff Srv  |
+ *
+ * API doc: If the Scheduler server is registered on the element then the Scene server shouldn't.
+ * The Scheduler model already includes the Scene model.
+ */
+static struct bt_mesh_model mock_models_elem[3] = {
+	{
+		.user_data = &sched_srv,
+		.id = BT_MESH_MODEL_ID_SCHEDULER_SRV,
+		.elem_idx = 0,
+	},
+	{
+		.user_data = &sched_cli,
+		.id = BT_MESH_MODEL_ID_SCHEDULER_CLI,
+		.elem_idx = 1,
+	},
+	{
+		.user_data = &onoff_srv,
+		.id = BT_MESH_MODEL_ID_GEN_ONOFF_SRV,
+		.elem_idx = 2,
+	},
+};
+
+static struct bt_mesh_elem mock_elem = {
+	.addr = 1,
+	.model_count = ARRAY_SIZE(mock_models_elem),
+	.models = mock_models_elem,
+};
+
+struct mock_comp_data {
+	size_t elem_count;
+	struct bt_mesh_elem *elem;
+} mock_comp = { .elem_count = 1, .elem = &mock_elem };
+
+static struct bt_mesh_model *mock_mod_srv = &mock_models_elem[0];
+static struct bt_mesh_model *mock_mod_cli = &mock_models_elem[1];
+
+static void ack_status_is_sent(bool val)
+{
+	ztest_check_expected_value(val);
+}
+
+/* Redefined mocks */
+
+struct bt_mesh_elem *bt_mesh_model_elem(struct bt_mesh_model *mod)
+{
+	return &mock_comp.elem[mod->elem_idx];
```

Comment:
Hmh. `mock_comp.elem` is a pointer to a single elem. `elem_idx` is defined up to 2. How does this even work? `elem_idx` should be 0 for all three models, right?